src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -400,6 +402,33 @@ public virtual async Task<ConnectionCompleteParams> Connect(ConnectParams connec
             return completeParams;
         }
 
+        private async Task<ConnectionCompleteParams> TryOpenConnectionWithRetry(ConnectionInfo connectionInfo, ConnectParams connectionParams)
+        {
+            int counter = 0;
+            ConnectionCompleteParams response = null;
+            while (counter <= MaxServerlessReconnectTries)
+            {
+                response = await TryOpenConnection(connectionInfo, connectionParams);
+                // If a serverless database is sleeping, it will return this error number and will need to be retried.
+                // See here for details: https://docs.microsoft.com/en-us/azure/azure-sql/database/serverless-tier-overview?view=azuresql#connectivity
+                if (response?.ErrorNumber == 40613)
+                {
+                    Logger.Write(TraceEventType.Information, "Attempt to retry serverless DB connection no. " + (counter + 1) + " for " + connectionInfo.OwnerUri);

```

Comment:
FYI - I added shortcut extension methods for all the logger stuff so you can just do `Logger.Information(...)`

And I'd suggest this for the message

`$"Database for connection {connectionInfo.OwnerUri} is paused, retrying connection. Attempt #{counter}"` (and move it below the counter++ (you'll also want to add an if check so this isn't printed out the last time when counter == MaxServerlessReconnectTries)